### PR TITLE
Convert the Transcript Line on the DOM to be a Custom Element.

### PR DESCRIPTION
**Pull request by amuamushu**
This change refactors the methods that involving a transcriptLine ul element by moving them inside the `TranscriptLine` Javascript class.

### CODE REVIEW

File: src/main/webapp/timestamps.js [hunk 1]
```diff
@@ -39,3 +39,13 @@ function timestampToString(timestampMs) {
 function secondsToMilliseconds(seconds) {
   return Math.round(seconds * MILLISECONDS_PER_SECOND);
 }
+
+/**
+ * Returns {String} a time range from
+ * `startTimestampMs` to `endTimestampMs`.
+ */
```
Comment (amuamushu) [inline src/main/webapp/timestamps.js:46]: This function is patched from pull request #228 . Please ignore this function for this pull request. Thank you!

File: src/main/webapp/view/transcript/transcript.js [hunk 2]
```diff
@@ -115,60 +86,108 @@ function seekTranscript(currentTime) {
...
- * is currently bolded.
- */
-function removeBold(transcriptLineLiElement) {
-  if (!isBolded(transcriptLineLiElement)) {
-    return;
+  /**
+   * Updates the template slot `slotName` with `slotValue` and styling
+   * from `slotStyle`.
+   */
+  updateSlot(slotName, slotValue, slotStyle) {
```
Comment (jaysaleh) [inline src/main/webapp/view/transcript/transcript.js:137]: nit: maybe rename this to `updateTemplateSlot`

Comment (amuamushu) [reply src/main/webapp/view/transcript/transcript.js:137]: done, thanks!

File: src/main/webapp/view/transcript/transcript.js [hunk 3]
```diff
@@ -115,60 +86,109 @@ function seekTranscript(currentTime) {
...
   }
-  transcriptLineLiElement.classList.add(DEFAULT_FONT_WEIGHT);
-  transcriptLineLiElement.classList.remove(BOLD_FONT_WEIGHT);
-}
 
-/** Returns true if`transcriptLineLiElement` is bolded. */
-function isBolded(transcriptLineLiElement) {
-  return transcriptLineLiElement.classList.contains(BOLD_FONT_WEIGHT);
-}
+  /** Returns true if the text is already bolded. */
```
Comment (n-wach) [inline src/main/webapp/view/transcript/transcript.js:146]: Change to `if this element is already bolded`?

File: src/main/webapp/view/transcript/transcript.js [hunk 4]
```diff
@@ -115,60 +86,109 @@ function seekTranscript(currentTime) {
...
- * is currently bolded.
- */
-function removeBold(transcriptLineLiElement) {
-  if (!isBolded(transcriptLineLiElement)) {
-    return;
+  /**
+   * Updates the template slot `slotName` with `slotValue` and styling
+   * from `slotStyle`.
+   */
+  updateTemplateSlot(slotName, slotValue, slotStyle) {
```
Comment (n-wach) [inline src/main/webapp/view/transcript/transcript.js:138]: Can the `slotStyle` be done in the template HTML code, and removed from this function?  Since it looks like it's always the same class per slot

File: src/main/webapp/view/transcript/transcript.js [hunk 5]
```diff
@@ -115,60 +86,109 @@ function seekTranscript(currentTime) {
...
+   * Removes bold from the text if it
+   * is currently bolded.
+   */
+  removeBold() {
+    if (!this.isBolded()) {
+      return;
+    }
+    this.classList.add(DEFAULT_FONT_WEIGHT);
+    this.classList.remove(BOLD_FONT_WEIGHT);
+  }
```
Comment (n-wach) [inline src/main/webapp/view/transcript/transcript.js:173]: Now that you're using classes, I feel like it would simplify things if `isBolded`, `addBold`, `removeBold`, were combined into a single `setBold(boolean)`.

Or even a `bold` setter and getter that adds/removes the styling and updates an internal state.

File: src/main/webapp/view/transcript/transcript.js [hunk 6]
```diff
@@ -115,60 +86,109 @@ function seekTranscript(currentTime) {
...
+    return this.startTimestampMs <= currentTimeMs &&
+        currentTimeMs <= this.endTimestampMs;
+  }
+
+  /**
+   * Scrolls this transcript line to the top of the transcript area.
+   */
+  scrollToTopOfTranscript() {
+    const transcriptContainer = document.getElementById(TRANSCRIPT_CONTAINER);
+    const ulElementOffset = this.parentElement.offsetTop;
```
Comment (n-wach) [inline src/main/webapp/view/transcript/transcript.js:189]: Is `this.parentElement` the same as `transcriptContainer`?

Comment (amuamushu) [reply src/main/webapp/view/transcript/transcript.js:189]: No, it's the ul tag that the transcript-line tags are nested under. `TranscriptContainer` contains the ul tag. The nesting of the tags is a bit weird though. I created an issue to fix this later (#236).

File: src/main/webapp/view/transcript/transcript.js (hunk 5 above)
Comment (n-wach) [reply src/main/webapp/view/transcript/transcript.js:171]: Ah, sounds good!

File: src/main/webapp/view/transcript/transcript.js (hunk 6 above)
Comment (n-wach) [reply src/main/webapp/view/transcript/transcript.js:187]: I was just wondering, nbd

File: src/main/webapp/view/transcript/transcript.js (hunk 4 above)
Comment (n-wach) [reply src/main/webapp/view/transcript/transcript.js:136]: Maybe you can wrap the `<slot>` in a `<span>` with the correct styling?  I know it's sometimes unavoidable but I think it's best practice to try to separate styling (especially static stuff like this) from code as much as possible

Comment (amuamushu) [reply src/main/webapp/view/transcript/transcript.js]: Gotcha, wrapping it works! Thanks!

File: src/main/webapp/view/transcript/transcript.js [hunk 7]
```diff
@@ -47,45 +51,13 @@ function addMultipleTranscriptLinesToDom(transcriptLines) {
...
-  // Sets the current transcript line to be the first line.
-  if (currentTranscriptLine == null) {
-    currentTranscriptLine = liElement;
-  }
-}
-
-/**
- * Creates a p tag to store the given `text` inside the
- * `container`.
+ * Creates a p tag to store the given {@code text} inside the
```
Comment (billie-cat) [inline src/main/webapp/view/transcript/transcript.js:59]: Should this @code be removed?

File: src/main/webapp/view/transcript/transcript.js [hunk 8]
```diff
@@ -113,60 +85,106 @@ function seekTranscript(currentTime) {
...
+  currentTranscriptLine.scrollToTopOfTranscript();
+  currentTranscriptLine.addBold();
   // TODO: Handle the case where the video isn't only playing.
 }
 
 /**
- * Bolds the text in `transcriptLineLiElement` if it is not already
- * bolded.
+ * Creates a transcript line element containing the text,
+ * start time, and end time from `transcriptLine`.
```
Comment (billie-cat) [inline src/main/webapp/view/transcript/transcript.js:101]: you are referencing transcriptLine but this is not related to the class. Please keep this generic for the class.

File: src/main/webapp/view/transcript/transcript.js [hunk 9]
```diff
@@ -113,60 +85,106 @@ function seekTranscript(currentTime) {
...
+   * Creates a custom HTML element representing a transcript line.
+   *
+   * <p>Uses the template and slots defined in `TRANSCRIPT_TEMPLATE` to
+   * help create the transcript line.
+   *
+   * @param transcriptLine The transcriptLine from `ENDPOINT_TRANSCRIPT`
+   *     whose `attributes` should be used.
+   */
+  constructor(transcriptLine) {
+    super();
```
Comment (billie-cat) [inline src/main/webapp/view/transcript/transcript.js:114]: It is best practise to not do work in the constructor. Please see: https://docs.google.com/document/d/14CEPF-d3eF6HzG4AhviQiRbdEaVF8_BDOwKHMUmqFEQ/edit

I would use a factory method to create the object.

Comment (n-wach) [reply src/main/webapp/view/transcript/transcript.js:97]: This is the same situation as in the DiscussionComment class, which we talked about and okay-ed.  Both use templates and need to initialize various things before they are rendered.

File: src/main/webapp/view/transcript/transcript.js [hunk 10]
```diff
@@ -113,60 +85,106 @@ function seekTranscript(currentTime) {
...
+    this.attachShadow({mode: 'open'});
+    this.shadowRoot.appendChild(template.content.cloneNode(true));
+    this.updateTemplateSlot(TRANSCRIPT_SLOT_TIME_RANGE, timestampRange);
+    this.updateTemplateSlot(TRANSCRIPT_SLOT_CONTENT, transcriptLine.content);
+
+    this.startTimestampMs = transcriptLine.startTimestampMs;
+    this.endTimestampMs = transcriptLine.endTimestampMs;
+    // Sets the current transcript line to be the first line.
+    if (currentTranscriptLine == null) {
+      currentTranscriptLine = this;
```
Comment (n-wach) [inline src/main/webapp/view/transcript/transcript.js:111]: This is an unclear side effect that should probably be moved out of the constructor

File: src/main/webapp/view/transcript/transcript.js [hunk 11]
```diff
@@ -113,60 +85,106 @@ function seekTranscript(currentTime) {
...
+        transcriptLine.startTimestampMs, transcriptLine.endTimestampMs);
+
+    const template = document.getElementById(TRANSCRIPT_TEMPLATE);
+    this.attachShadow({mode: 'open'});
+    this.shadowRoot.appendChild(template.content.cloneNode(true));
+    this.updateTemplateSlot(TRANSCRIPT_SLOT_TIME_RANGE, timestampRange);
+    this.updateTemplateSlot(TRANSCRIPT_SLOT_CONTENT, transcriptLine.content);
+
+    this.startTimestampMs = transcriptLine.startTimestampMs;
+    this.endTimestampMs = transcriptLine.endTimestampMs;
```
Comment (n-wach) [inline src/main/webapp/view/transcript/transcript.js:108]: It might be better to just save everything with `this.transcriptLine = transcriptLine` instead of copying over the various values you'll need later.

File: src/main/webapp/view/transcript/transcript.js [hunk 12]
```diff
@@ -113,60 +85,106 @@ function seekTranscript(currentTime) {
...
+    }
+    this.classList.add(DEFAULT_FONT_WEIGHT);
+    this.classList.remove(BOLD_FONT_WEIGHT);
+  }
+
+  /**
+   * Returns true if `currentTimeMs` is within the time range for
+   * this transcript line.
+   */
+  isWithinCurrentTimeRange(currentTimeMs) {
```
Comment (n-wach) [inline src/main/webapp/view/transcript/transcript.js:158]: Can you change method and argument name to be more generic?

File: src/main/webapp/view/transcript/transcript.js (hunk 10 above)
Comment (amuamushu) [reply src/main/webapp/view/transcript/transcript.js]: True, I deleted this if statement for now and added a TODO for making this an instance attribute when `transcript.js` is refactored into a class. This should move the responsibility of keeping track of the current transcript line outside of the `transcriptLine` class and to where `seekTranscript()` is. Thanks!

File: src/main/webapp/view/transcript/transcript.js (hunk 12 above)
Comment (amuamushu) [reply src/main/webapp/view/transcript/transcript.js]: Thanks for catching that! I must have missed it in this branch.

File: src/main/webapp/view/transcript/transcript.js (hunk 8 above)
Comment (amuamushu) [reply src/main/webapp/view/transcript/transcript.js]: gotcha, I remove the last bit (`from transcriptLine`) from the description.

File: src/main/webapp/view/transcript/transcript.js (hunk 7 above)
Comment (amuamushu) [reply src/main/webapp/view/transcript/transcript.js]: Yes, thank you for catching that!

File: src/main/webapp/view/transcript/transcript.js (hunk 9 above)
Comment (billie-cat) [reply src/main/webapp/view/transcript/transcript.js:99]: I ended up talking to my co-worker about this and independent of the language, doing work in the constructor is not a good practise. It is best to create a static method in the class which does all the work and passes the fields to the constructor.

File: src/main/webapp/view/transcript/transcript.js [hunk 13]
```diff
@@ -13,12 +13,18 @@
...
 const DEFAULT_FONT_WEIGHT = 'text-muted';
 const BOLD_FONT_WEIGHT = 'font-weight-bold';
-const PARAM_ID = 'id';
+const URL_PARAM_ID = 'id';
+const TRANSCRIPT_SLOT_TIME_RANGE = 'timestamp-range';
+const TRANSCRIPT_SLOT_CONTENT = 'content';
+const CUSTOM_ELEMENT_TRANSCRIPT_LINE = 'transcript-line';
 
 let /** Element */ currentTranscriptLine;
+// TODO: Create an instance reference to the current transcriptLine when
```
Comment (billie-cat) [inline src/main/webapp/view/transcript/transcript.js:26]: should "current transcriptLine be currentTranscriptLine?

File: src/main/webapp/view/transcript/transcript.js [hunk 14]
```diff
@@ -113,60 +70,101 @@ function seekTranscript(currentTime) {
...
 /**
- * Bolds the text in `transcriptLineLiElement` if it is not already
- * bolded.
+ * Creates a transcript line element containing the text,
+ * start time, and end time.
  */
-function addBold(transcriptLineLiElement) {
-  if (isBolded(transcriptLineLiElement)) {
-    return;
+class TranscriptLine extends HTMLElement {
```
Comment (billie-cat) [inline src/main/webapp/view/transcript/transcript.js:88]: TranscriptLine is slightly generic class name? What will your other class names be for TranscriptLine related code? I thought you were going to name this TranscriptLineElement and then have another class be TranscriptLineSeeker (or something like that).

File: src/main/webapp/view/transcript/transcript.js [hunk 15]
```diff
@@ -113,60 +70,101 @@ function seekTranscript(currentTime) {
...
+  /**
+   * Creates a custom HTML element representing a transcript line.
```
Comment (billie-cat) [inline src/main/webapp/view/transcript/transcript.js:90]:
```suggestion
   * Creates a custom HTML element representing `transcriptLine`.
```

File: src/main/webapp/view/transcript/transcript.js [hunk 16]
```diff
@@ -113,60 +70,101 @@ function seekTranscript(currentTime) {
...
- * Returns true if `currentTimeMs` is within the time range for
- * the current transcript line.
- */
-function isWithinCurrentTimeRange(currentTimeMs) {
-  return currentTranscriptLine.startTimestampMs <= currentTimeMs &&
-      currentTimeMs <= currentTranscriptLine.endTimestampMs;
-}
+  /**
+   * Bolds the element if it is not already
+   * bolded.
```
Comment (billie-cat) [inline src/main/webapp/view/transcript/transcript.js:129]: nit: the new line for the work bolded seems a bit weird. You should only add new lines if it goes above a certain character length.

File: src/main/webapp/view/transcript/transcript.js [hunk 17]
```diff
@@ -113,60 +70,101 @@ function seekTranscript(currentTime) {
...
+    }
+    this.classList.add(DEFAULT_FONT_WEIGHT);
+    this.classList.remove(BOLD_FONT_WEIGHT);
+  }
+
+  /**
+   * Returns true if `timeMs` is within the time range for
+   * this transcript line.
+   */
+  isWithinTimeRange(timeMs) {
```
Comment (billie-cat) [inline src/main/webapp/view/transcript/transcript.js:155]: nit: is timeMs a timestamp? should it be timestampMs?

File: src/main/webapp/view/transcript/transcript.js [hunk 18]
```diff
@@ -113,60 +70,101 @@ function seekTranscript(currentTime) {
...
+    return this.transcriptLine.startTimestampMs <= timeMs &&
+    timeMs <= this.transcriptLine.endTimestampMs;
+  }
+
+  /**
+   * Scrolls this transcript line to the top of the transcript area.
+   */
+  scrollToTopOfTranscript() {
```
Comment (billie-cat) [inline src/main/webapp/view/transcript/transcript.js:163]: if this class is about a TranscriptLineElement, then how can it scroll to the top of the transcript? That would implement multiple transcript lines.

File: src/main/webapp/view/transcript/transcript.js (hunk 9 above)
Comment (f-camp) [reply src/main/webapp/view/transcript/transcript.js:111]: The part I have a problem with in this ```createTimestampRange()``` That seems to me the part that is doing some work. If we can remove that and put it in some kind of initialization method that would work for me.

Comment (amuamushu) [reply src/main/webapp/view/transcript/transcript.js]: (Resolved in chat) I kept the code for attaching the shadow in the constructor but moved the code for creating the timerange and updating the slots to a static method.

File: src/main/webapp/timestamps.js (hunk 1 above)
Comment (billie-cat) [reply src/main/webapp/timestamps.js:46]: Will you remove this before merging this PR?

Comment (amuamushu) [reply src/main/webapp/timestamps.js:46]: Yes :D
